File: ports/php/test/stream_safe_violation_test.php
<?php

declare(strict_types=1);

require_once __DIR__ . '/test_helper.php';

use UnicodePhp\Security\Form\StreamSafeViolation;
use UnicodePhp\Security\Mode;
use UnicodePhp\Security\Policy;
use UnicodePhp\Security\Profile;

/** @param list<int> $expected @param list<int> $actual */
function assert_stream_positions(array $expected, array $actual, string $label): void
{
    assert_same_value($expected, $actual, $label);
}

/**
 * "a" followed by $n combining acute accents (U+0301, CCC = 230).
 *
 * @return list<int>
 */
function a_plus_marks(int $n): array
{
    $v = [0x61];
    for ($i = 0; $i < $n; $i++) {
        $v[] = 0x0301;
    }
    return $v;
}

// ── (a) shared context-free fixture, driven through scanForms ───────────────
$fixture = fixture_json('detectors/stream_safe_violation.json');
foreach ($fixture['cases'] as $case) {
    $verdict = Policy::scanForms(Profile::OpaqueSecret, Mode::Observe, $case['input']);
    $codes = verdict_codes($verdict);
    foreach ($case['required_findings'] as $required) {
        assert_includes_value($codes, $required, 'fixture/' . $case['name']);
    }
    if ($case['required_findings'] === []) {
        $needle = '.' . $fixture['family'] . '.';
        foreach ($codes as $code) {
            if (str_contains($code, $needle)) {
                throw new RuntimeException('fixture/' . $case['name'] . ' unexpected family ' . $fixture['family']);
            }
        }
    }
}

// ── (b) every rust `#[test]` vector transcribed verbatim ────────────────────
$vectors = 0;

// detect_empty_clear
$v = StreamSafeViolation::detect([]);
assert_same_value(true, $v->classify->isClear(), 'empty clear');
assert_same_value(null, $v->classify->tag(), 'empty tag none');
assert_same_value(0, $v->maxRunLen, 'empty maxRunLen');
assert_same_value(0, $v->overrunCount, 'empty overrunCount');
assert_same_value(0, $v->totalNonStarters, 'empty totalNonStarters');
$vectors++;

// detect_ascii_clear — "Hello"
$v = StreamSafeViolation::detect([0x48, 0x65, 0x6C, 0x6C, 0x6F]);
assert_same_value(true, $v->classify->isClear(), 'ascii clear');
assert_same_value(0, $v->maxRunLen, 'ascii maxRunLen');
assert_same_value(0, $v->totalNonStarters, 'ascii totalNonStarters');
$vectors++;

// detect_one_combine_clear — "a" + U+0301
$v = StreamSafeViolation::detect([0x61, 0x0301]);
assert_same_value(true, $v->classify->isClear(), 'one combine clear');
assert_same_value(1, $v->maxRunLen, 'one combine maxRunLen');
assert_same_value(0, $v->overrunCount, 'one combine overrunCount');
assert_same_value(1, $v->totalNonStarters, 'one combine totalNonStarters');
$vectors++;

// detect_thirty_marks_clear — boundary case, stays clear under strict `>`
$v = StreamSafeViolation::detect(a_plus_marks(30));
assert_same_value(true, $v->classify->isClear(), 'thirty marks clear');
assert_same_value(null, $v->classify->tag(), 'thirty marks tag none');
assert_same_value(30, $v->maxRunLen, 'thirty marks maxRunLen');
assert_same_value(0, $v->overrunCount, 'thirty marks overrunCount');
assert_same_value(30, $v->totalNonStarters, 'thirty marks totalNonStarters');
$vectors++;

// detect_thirtyone_marks_hazard — fires with firstOverrun = (1, 31), positions [1]
$v = StreamSafeViolation::detect(a_plus_marks(31));
assert_same_value(false, $v->classify->isClear(), 'thirtyone marks hazard');
assert_same_value('StreamSafeOverrun', $v->classify->tag(), 'thirtyone marks tag');
assert_stream_positions([1], $v->classify->positions(), 'thirtyone marks positions');
assert_same_value(1, $v->classify->sub->basePos, 'thirtyone marks basePos');
assert_same_value(31, $v->classify->sub->runLen, 'thirtyone marks runLen');
assert_same_value(31, $v->maxRunLen, 'thirtyone marks maxRunLen');
assert_same_value(1, $v->overrunCount, 'thirtyone marks overrunCount');
assert_same_value(31, $v->totalNonStarters, 'thirtyone marks totalNonStarters');
$vectors++;

// bare_mark_run_starts_at_zero — 31 bare marks, run opens at index 0
$bare = [];
for ($i = 0; $i < 31; $i++) {
    $bare[] = 0x0301;
}
$v = StreamSafeViolation::detect($bare);
assert_same_value('StreamSafeOverrun', $v->classify->tag(), 'bare run tag');
assert_stream_positions([0], $v->classify->positions(), 'bare run positions');
assert_same_value(31, $v->maxRunLen, 'bare run maxRunLen');
assert_same_value(31, $v->totalNonStarters, 'bare run totalNonStarters');
$vectors++;

// two_short_runs_clear_totals_summed — "a" + 30 marks + "b" + 30 marks
$two = a_plus_marks(30);
$two[] = 0x62;
for ($i = 0; $i < 30; $i++) {
    $two[] = 0x0301;
}
$v = StreamSafeViolation::detect($two);
assert_same_value(true, $v->classify->isClear(), 'two short runs clear');
assert_same_value(30, $v->maxRunLen, 'two short runs maxRunLen');
assert_same_value(0, $v->overrunCount, 'two short runs overrunCount');
assert_same_value(60, $v->totalNonStarters, 'two short runs totalNonStarters');
$vectors++;

// first_overrun_reports_long_run_start — "a" + 5 marks + "b" + 31 marks
$first = a_plus_marks(5);
$first[] = 0x62;
for ($i = 0; $i < 31; $i++) {
    $first[] = 0x0301;
}
$v = StreamSafeViolation::detect($first);
assert_same_value('StreamSafeOverrun', $v->classify->tag(), 'first overrun tag');
assert_stream_positions([7], $v->classify->positions(), 'first overrun positions');
assert_same_value(31, $v->maxRunLen, 'first overrun maxRunLen');
assert_same_value(1, $v->overrunCount, 'first overrun overrunCount');
assert_same_value(36, $v->totalNonStarters, 'first overrun totalNonStarters');
$vectors++;

// ── (c) 30/31 boundary drives the reason code through the policy scan ───────
$clear = Policy::scanForms(Profile::OpaqueSecret, Mode::Observe, a_plus_marks(30));
foreach (verdict_codes($clear) as $code) {
    if (str_contains($code, '.stream-safe-violation.')) {
        throw new RuntimeException('30-mark boundary must stay clear, got ' . $code);
    }
}
$hazard = Policy::scanForms(Profile::OpaqueSecret, Mode::Observe, a_plus_marks(31));
assert_includes_value(
    verdict_codes($hazard),
    'unicode.security.F.stream-safe-violation.StreamSafeOverrun',
    '31-mark boundary reason code',
);

if ($vectors !== 8) {
    throw new RuntimeException('expected 8 rust vectors, transcribed ' . $vectors);
}

echo "stream_safe_violation_test: OK (" . $vectors . " rust vectors)\n";

File: ports/lua/test/stream_safe_violation_test.lua
local h = require("test_helper")
local calculus = require("unicode_lua.security.calculus")
local policy = require("unicode_lua.security.policy")
local ss = require("unicode_lua.security.form.stream_safe_violation")

local Family = calculus.Family

-- Map a detect verdict to the reason codes it would contribute, using the same
-- family/layer wiring the policy engine applies (F2 ⇒
-- unicode.security.F.stream-safe-violation.<sub>).
local function codes_of(verdict)
  local out = {}
  if not ss.classification_is_clear(verdict.classify) then
    out[#out + 1] = policy.reason_code(Family.StreamSafeViolation, ss.classification_tag(verdict.classify))
  end
  return out
end

-- ── The reason-code wiring is exactly the fixture's required finding ──────────
h.assert_equal(
  policy.reason_code(Family.StreamSafeViolation, "StreamSafeOverrun"),
  "unicode.security.F.stream-safe-violation.StreamSafeOverrun",
  "reason-code wiring"
)

-- ── Shared fixture through detect ─────────────────────────────────────────────
local ROOT = "testdata"
local fixture = h.json_file(ROOT .. "/fixtures/security/detectors/stream_safe_violation.json")
h.assert_equal(fixture.family, "stream-safe-violation", "fixture family")

for _, case in ipairs(fixture.cases) do
  local verdict = ss.detect(case.input)
  local codes = codes_of(verdict)
  for _, required in ipairs(case.required_findings) do
    h.assert_includes(codes, required, "fixture/" .. case.name)
  end
  if #case.required_findings == 0 then
    for _, code in ipairs(codes) do
      if code:find("%." .. fixture.family .. "%.", 1) ~= nil then
        error("fixture/" .. case.name .. " unexpected family " .. fixture.family)
      end
    end
  end
end

-- ── 30 / 31 boundary (strict `>` at STREAM_SAFE_LIMIT = 30) ───────────────────
local ACUTE = 0x0301 -- COMBINING ACUTE ACCENT, CCC = 230 (a non-starter)

local function a_plus_marks(n)
  local v = { 0x61 }
  for _ = 1, n do
    v[#v + 1] = ACUTE
  end
  return v
end

h.assert_equal(ss.STREAM_SAFE_LIMIT, 30, "stream-safe limit")

-- Exactly 30 marks after a starter is the boundary case — stays clear under `>`.
local thirty = ss.detect(a_plus_marks(30))
if not ss.classification_is_clear(thirty.classify) then
  error("30-mark boundary should be clear")
end
h.assert_equal(ss.classification_tag(thirty.classify), nil, "30-mark tag nil")
h.assert_equal(thirty.max_run_len, 30, "30-mark max_run_len")
h.assert_equal(thirty.overrun_count, 0, "30-mark overrun_count")
h.assert_equal(thirty.total_non_starters, 30, "30-mark total_non_starters")

-- 31 marks after a starter fires StreamSafeOverrun with base_pos 1, run_len 31.
local thirtyone = ss.detect(a_plus_marks(31))
if ss.classification_is_clear(thirtyone.classify) then
  error("31-mark boundary should be a hazard")
end
h.assert_equal(ss.classification_tag(thirtyone.classify), "StreamSafeOverrun", "31-mark tag")
h.assert_equal(ss.classification_positions(thirtyone.classify), { 1 }, "31-mark positions")
h.assert_equal(thirtyone.classify.sub.base_pos, 1, "31-mark base_pos")
h.assert_equal(thirtyone.classify.sub.run_len, 31, "31-mark run_len")
h.assert_equal(thirtyone.max_run_len, 31, "31-mark max_run_len")
h.assert_equal(thirtyone.overrun_count, 1, "31-mark overrun_count")
h.assert_equal(thirtyone.total_non_starters, 31, "31-mark total_non_starters")

-- ── Run-inventory structure (mirrors the reference's collectRunsGo tests) ─────

-- A bare non-starter run opening at index 0 records its start as 0.
local bare = {}
for _ = 1, 31 do
  bare[#bare + 1] = ACUTE
end
local bare_v = ss.detect(bare)
h.assert_equal(ss.classification_tag(bare_v.classify), "StreamSafeOverrun", "bare tag")
h.assert_equal(ss.classification_positions(bare_v.classify), { 0 }, "bare positions")
h.assert_equal(bare_v.max_run_len, 31, "bare max_run_len")
h.assert_equal(bare_v.total_non_starters, 31, "bare total_non_starters")

-- Two separate runs, each under the limit, stay clear but sum into the totals.
local two = a_plus_marks(30)
two[#two + 1] = 0x62
for _ = 1, 30 do
  two[#two + 1] = ACUTE
end
local two_v = ss.detect(two)
if not ss.classification_is_clear(two_v.classify) then
  error("two short runs should be clear")
end
h.assert_equal(two_v.max_run_len, 30, "two-run max_run_len")
h.assert_equal(two_v.overrun_count, 0, "two-run overrun_count")
h.assert_equal(two_v.total_non_starters, 60, "two-run total_non_starters")

-- The first overrun wins: a short run preceding a long run does not shadow it,
-- and the reported base_pos is the long run's start. "a" + 5 + "b" + 31 marks.
local first = a_plus_marks(5)
first[#first + 1] = 0x62
for _ = 1, 31 do
  first[#first + 1] = ACUTE
end
local first_v = ss.detect(first)
h.assert_equal(ss.classification_tag(first_v.classify), "StreamSafeOverrun", "first-overrun tag")
h.assert_equal(ss.classification_positions(first_v.classify), { 7 }, "first-overrun positions")
h.assert_equal(first_v.max_run_len, 31, "first-overrun max_run_len")
h.assert_equal(first_v.overrun_count, 1, "first-overrun overrun_count")
h.assert_equal(first_v.total_non_starters, 36, "first-overrun total_non_starters")
